### ISSUE TITLE
Feature Request: Develop a better CI process that performs better testing of all functionality

### ISSUE THREAD

**Issue by abraunegg** [abraunegg/onedrive#711]
As the title suggests, whilst there is a Travis CI script that automates the build testing, and performs limited upload / download testing, there needs to be a more robust test suite to test / validate all functions & combinations to avoid simple regressions and the like.

Would love some assistance here in this regards.

**Comment by K1ngjulien** [comment]
I'd love to help. What can I do?

**Comment by abraunegg** [comment]
@K1ngjulien 
Many thanks for responding :)

Essentially I am 100% open for discussion here. Currently, this is what gets tested via Travis CI - https://api.travis-ci.com/v3/job/262744734/log.txt

My thoughts:

**Assumptions:**
* Application binary has been built. Existing Travis CI process tests this
* Application has been authorised - valid configuration exists in ~/.config/onedrive/
* Anyone can run the 'tests' using their own 'test' OneDrive account

**Thoughts:**
* Some sort of script (that can be run without adding a toolkit or something major) that runs through basic functionality
* Tests functionality permutations of these individually, together, odd pairing
  * --upload-only
  * --download-only
  * sync_list
  * skip_dir
  * skip_file
  * skipping files / folders based on .nosync
  * symbolic links
  * (basically all options that can be configured)
* Issuing "back channel" API calls to OneDrive to simulate moving files / uploading files via WebUI
* Easyish to maintain / add / modify test cases when functionality is added / modified / improved

**Goals:**
* Simplify testing when code changes are done to ensure that regressions are avoided as much as possible


Hopefully that makes some sense :)

**Comment by K1ngjulien** [comment]
what do you mean by "simulate via WebUI"? Adding a Script controlled browser with something like puppeteer?

Other than that, this seems like a good plan :)

How are you getting the onedrive refresh token in the ci? Is it an environment variable?

**Comment by abraunegg** [comment]
@K1ngjulien 
> what do you mean by "simulate via WebUI"? Adding a Script controlled browser with something like puppeteer?

Open to suggestions. I was thinking of some sort of python script or similar that simply called the 'move' or whatever API directly if possible.

> How are you getting the onedrive refresh token in the ci? Is it an environment variable?

For Travis CI, yes this is a project session variable, but I was thinking that whatever this is turned into, anyone can run with their user credentials being set if required, so not just limited to this project / Travis CI repository if that makes sense.

**Comment by K1ngjulien** [comment]
Ok using the pyhton sdk should be pretty easy and I could implement that. But saving an OAuth session creates a .pickle file. Is there a way to put that file into the CI without commiting it to git?
I have never used Travis, only Gitlab, so please excuse the noob questions ^^

**Comment by abraunegg** [comment]
@K1ngjulien 
The assumption I was running / thinking would be that the application is already authenticated / configured. If you wanted to also test the auth process & logout etc - up to you :)

For Travis CI, it can pull in the auth as a variable (hidden of course) which is how I do things at present.

**Comment by K1ngjulien** [comment]
I don't think I quite understand ^^
Don't you need to authenticate the python script as well in order to move the files?

**Comment by abraunegg** [comment]
@abraunegg 
If the application is already 'authenticated' there is the oauth token in ~/.config/onedrive/ .. so my thoughts were that if this was there / present - that could be reused rather than having to re-auth

If reusing, then calling the API directly would be possible I was thinking

**Comment by K1ngjulien** [comment]
@abraunegg 
ahh I understand now. saves on having to store two auth tokens.

regardless of that, I think we should make the testing a bit more modular so that it will be easier to add new tests in the future.
I drew a quick diagram of how I think this could look, but I don't have enough experience to say if doing it this way will be a good idea.

![ci_onedrive](https://user-images.githubusercontent.com/16562333/70389349-a91dbb00-19be-11ea-9314-8b47edf7ae0f.png)

start_ci.sh would detect which arch we are on and run the corresponding build script (which may just be a different make call).
those would then call run_tests.sh which in turn would loop over all test scripts to run all the tests.
the tests in turn could run python scripts if needed or do something else, idk.

**Comment by abraunegg** [comment]
@K1ngjulien 
Seems like a reasonable idea.

The trick would be also to ensure that the script is checking for the response codes to know that the operation was successful or failed, or at least looking at the logging going on.

This way, you could do something like this:
```
Testing file upload [success]
Testing folder upload [success]
Testing bad file sequences [failed - test 4]
```

Thoughts?

**Comment by K1ngjulien** [comment]
@abraunegg 
Looks good and seems easy to do. When a test fails we could either still try to run all the others or cancel right there to save on time.

I'd like to set up travis on my forked repo so I can try to implement some of this. Can you help me with that?
Where exactly to I put the Access token? I should probably use a brand new onedrive account, right?

**Comment by abraunegg** [comment]
@K1ngjulien 
> I'd like to set up travis on my forked repo so I can try to implement some of this. Can you help me with that?

I think that the easiest option here I think would be to get the scripts working in a 'manual' fashion locally first, then bring in those changes as a PR modifying https://github.com/abraunegg/onedrive/blob/master/.travis-ci.sh so that the `test_onedrive` function performs the testing.

Via a PR, I can then check out the PR, and run the tests manually as well against a test account to validate etc.

If you want to test this yourself via Travis CI, essentially there is a bit of work to do - integrate your account into Travis CI, configure access, and then have Travis run the above shell script to build / test your work.

Right now, I use a separate OneDrive Test Account, but only perform functional testing on the x86_64 build, as what happens is that Travis CI is building / testing in parallel, therefor you get 'bad results' when you have 4 different architectures all try to upload / download test data from the same account at the same time. The fix for that is to use 4 test accounts (1 per arch) but also pointless as the functionality should not be different depending on the architecture of the platform.

In testing, I also use the following scripts to make random files:
**files:**
```
#!/bin/bash
FILECOUNT=0
DIRCOUNT=0
DESTDIR="/home/alex/OneDrive/random_files"
FILES=10
FOLDERS=5

mkdir -p $DESTDIR

while [ $DIRCOUNT -lt $FOLDERS ] ; do
        DIRNAME=`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1`
        DIRPATH=$DESTDIR/$DIRNAME
        mkdir -p $DIRPATH
        while [ $FILECOUNT -lt $FILES ] ; do
                FILESIZE=$RANDOM
                FILESIZE=$(($FILESIZE+1024))
                FILENAME=$DIRPATH/file${FILECOUNT}.data
                echo "Creating $FILENAME"
                dd if=/dev/urandom of=${FILENAME} bs=256 count=$FILESIZE > /dev/null 2>&1
                ((FILECOUNT++))
        done
        FILECOUNT=0
        ((DIRCOUNT++))
done
```

**images:**
```
#!/bin/bash
FILECOUNT=0
DIRCOUNT=0
DESTDIR="/home/alex/OneDrive/random_images"
FILES=50
FOLDERS=1

mkdir -p $DESTDIR

while [ $DIRCOUNT -lt $FOLDERS ] ; do
        DIRNAME=`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1`
        DIRPATH=$DESTDIR/$DIRNAME
        mkdir -p $DIRPATH
        while [ $FILECOUNT -lt $FILES ] ; do
                FILENAME=$DIRPATH/image${FILECOUNT}.png
                echo "Creating $FILENAME"
                mx=320;my=240;head -c "$((3*mx*my))" /dev/urandom | convert -depth 8 -size "${mx}x${my}" RGB:- ${FILENAME}
                ((FILECOUNT++))
        done
        FILECOUNT=0
        ((DIRCOUNT++))
done
```

**video:**
```
#!/bin/bash
FILECOUNT=0
DIRCOUNT=0
DESTDIR="/home/alex/OneDrive/random_videos"
FILES=5
FOLDERS=1

mkdir -p $DESTDIR

while [ $DIRCOUNT -lt $FOLDERS ] ; do
        DIRNAME=`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1`
        DIRPATH=$DESTDIR/$DIRNAME
        mkdir -p $DIRPATH
        while [ $FILECOUNT -lt $FILES ] ; do
                FILENAME=$DIRPATH/test_video_${FILECOUNT}.mpg
                # Duration is number of seconds the video will run for
                DURATION=$(shuf -i 30-60 -n 1)
                echo "Creating $FILENAME"
                ffmpeg -f lavfi -i testsrc=duration=$DURATION:size=1280x720:rate=30 $FILENAME
                ((FILECOUNT++))
        done
        FILECOUNT=0
        ((DIRCOUNT++))
done
```

The use of these scripts would require the additional dependencies to be loaded (if used via Travis CI)

**Comment by abraunegg** [comment]
@K1ngjulien
Any update on this ?

**Comment by K1ngjulien** [comment]
Sorry I've been busy with school, so I haven't had time to continue working on this.

**Comment by abraunegg** [comment]
@K1ngjulien 
Any update in making any progress here in the last 3 months?

**Comment by kapec94** [comment]
Hi there,
let me just make some input in regard of testing:

1. Beside of having high-level scripting tests that operate on "`onedrive` binary -- actual OneDrive server" level, code should be strongly unit-tested to ensure its proper operation in a reliable and extensible way. The code base is pretty massive with lots of enormous functions that may have lots of quirks and even a small change could introduce unexpected regressions. Unit-testing would be only limited to checking workings of the code, with OneDrive API and even SQLite DB mocked (you tell the mocks what to expect and what to return, application code interacts with those mocks not even knowing that those are mocks instead of real-life services).
2. GitHub now supports [GitHub Actions](https://github.com/features/actions) CI/CD thingy that for public repositories has unlimited processing time, so this could be leveraged to improve integration and testing of `onedrive`.
3. A good starting point would be to have a suite of manual tests defined, so the testing can be more streamlined without much additional coding. This benefits everyone: @abraunegg as the maintainer is less likely to forget testing some quirky behavior and other contributors have an idea what and how should be tested and even how big is the scope of tests to be performed if some change is made.
4. For testing with shell scripts I recommend [bats](https://github.com/sstephenson/bats).
5. I suggest that for delegating tasks, even as kind of R&D thing, as @K1ngjulien proposed he'll try out some things, there should be a created separate ticket, so the asignee could pestered at one's pleasure. :smiley:

**Comment by abraunegg** [comment]
Closing this issue as current CI testing will have to be sufficient

**Comment by kvalev** [comment]
I think there is still one aspect, which can be improved - the automated build and publish of arm images. However given that this is not currently possible using the DockerHub hooks, the only viable solution is to switch to a different CI system, for example GitHub actions.
